mistralrs-core/src/vision_models/llava/llava15.rs

```diff
@@ -0,0 +1,333 @@
+                    &(..),
+                ],
+                &image_features_vec[(image_id - 1) as usize],
+            )?;
+        }
+        //truncate
+        let (_, seq_len) = input_ids.shape().dims2()?;
+        if seq_len > self.config.text_config.max_length {
+            result = result.i((.., ..self.config.text_config.max_length, ..))?
+        }
+        Ok(result)
+    }
+
+    pub fn forward_inputs(
+        &self,
+        input_ids: &Tensor,
+        pixel_values: Option<Tensor>,
+        num_image_tokens: Option<usize>,
+        seqlen_offsets: &[usize],
+        start_offsets_kernel: Tensor,
+        context_lens: Vec<(usize, usize)>,
+        position_ids: Vec<usize>,
+    ) -> Result<Tensor> {
+        if let Some(ref pixel_values) = pixel_values {
+            // we assume(as it should be) only prompt request contains image
+            let input_embeds = self.prepare_inputs_labels_for_multimodal(
+                input_ids,
+                pixel_values,
+                num_image_tokens.unwrap(),
+            )?;
+            self.llm.forward_input_embed(
+                input_ids,
+                input_embeds,
+                seqlen_offsets,
+                start_offsets_kernel,
+                context_lens,
+            )
+        } else {
+            self.llm.forward(
+                input_ids,
+                seqlen_offsets,
+                start_offsets_kernel,
+                context_lens,
+                position_ids,
+            )
+        }
+    }
+}
+
+impl IsqModel for Model {
+    fn get_tensors(&mut self) -> (Vec<(&mut QMatMul, Option<usize>)>, &dyn DeviceMapper) {
+        // I don't really get this part
+        self.llm.get_tensors()
+    }
+}
+
+impl VisionModel for Model {
+    fn forward(
+        &self,
+        input_ids: &Tensor,
+        pixel_values: Option<Tensor>,
+        seqlen_offsets: &[usize],
+        start_offsets_kernel: Tensor,
+        context_lens: Vec<(usize, usize)>,
+        position_ids: Vec<usize>,
+        _model_specific_args: Box<dyn std::any::Any>, // pixel attention mask, or image sizes, or anything else
+    ) -> candle_core::Result<Tensor> {
+        self.forward_inputs(
+            input_ids,
+            pixel_values,
+            Some(
+                self.clip_vision_tower.num_patches_per_side()
+                    * self.clip_vision_tower.num_patches_per_side(),
+            ),
+            seqlen_offsets,
+            start_offsets_kernel,
+            context_lens,
+            position_ids,
+        )
+    }
+
+    fn device(&self) -> &Device {
+        &self.device
+    }
+
+    fn cache(&self) -> &crate::pipeline::Cache {
+        self.llm.cache()
+    }
+
+    fn max_seq_len(&self) -> usize {
+        self.config.text_config.max_length
+    }
+
+    fn has_conv2d(&self) -> bool {
+        true
+    }
+}
+
+impl AnyMoeBaseModelMixin for Model {
+    //untested
```

Comment:
That's fine, given it forwards the methods to tested models, if you could remove this that would be great :)

Reply:
removed